### PR TITLE
Added links to workbench

### PR DESCRIPTION
**Pull request by SyntaxXeror**
Just a quick PR for feedback on the workbench.

ATTENTION: For now, there is no test.

### CODE REVIEW

**Review by prasadtalasila — CHANGES_REQUESTED** [review]

File: client/src/route/workbench/WorkbenchButtonData.tsx [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    icon: <CodeOutlinedIcon />,
+  },
+  {
+    name: 'JupyterLab',
+    endpoint: '',
+    icon: <ScienceOutlinedIcon />,
+  },
+  // Add more buttons as needed
+];
+
```
[the remark this answers is not in the document]
Comment (SyntaxXeror) [reply client/src/route/workbench/WorkbenchButtonData.tsx:30]: Yes, do you want the actual Jupyter icon for the button? 
We found out that VSCode doesn't allow the use of their icon in the way we use it, so we chose not to go for the actual icons for both VSCode and Jupyter.

[the remark this answers is not in the document]
Comment (prasadtalasila) [reply client/src/route/workbench/WorkbenchButtonData.tsx:30]: Nope, the current ones look good.

File: client/src/components/LinkButtons.tsx [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+interface LinkButtonsProps {
+  buttons: Button[];
+  link?: string;
+}
+
+const StyledIconButton = styled(IconButton)`
+  margin: 8px;
+`;
+
+const IconLabel = styled.div`
```
[the remark this answers is not in the document]
Comment (SyntaxXeror) [reply client/src/components/LinkButtons.tsx:22]: @prasadtalasila I'm puzzled by what you mean by this comment. I have reduced one div maybe two, but reducing more alters the design a lot.

File: client/src/route/workbench/WorkbenchButtonData.tsx [hunk 3]
```diff
@@ -0,0 +1,31 @@
+import * as React from 'react';
+import TerminalOutlinedIcon from '@mui/icons-material/TerminalOutlined';
+import DesktopWindowsOutlinedIcon from '@mui/icons-material/DesktopWindowsOutlined';
+import CodeOutlinedIcon from '@mui/icons-material/CodeOutlined';
+import ScienceOutlinedIcon from '@mui/icons-material/ScienceOutlined';
+
+const buttons = [
```
[the remark this answers is not in the document]
Comment (SyntaxXeror) [reply client/src/route/workbench/WorkbenchButtonData.tsx:7]: @prasadtalasila placing the data (array) in envUtil.ts will just give us the problem with code climate where the function then exceeds 25 lines of code. Then taking the data and placing it in envUtil.ts but outside the function, wouldn't that take the smart part out of placing the array in envUtil.ts? We propose we either keep the data adjacent to the route, where it's needed, or (reluctantly) keep the data inside envUtil.ts(x)

File: client/src/components/LinkButtons.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (prasadtalasila) [reply client/src/components/LinkButtons.tsx:22]: Ok then, reduce as much as you can and leave the rest.

File: client/src/route/workbench/WorkbenchButtonData.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (prasadtalasila) [reply client/src/route/workbench/WorkbenchButtonData.tsx:7]: Can you place a commit that would show the code giving codeclimate problems? 
I'm also reluctantly proposing the refactoring of routes to envUtils.ts (strictly no JSX in this file). But am concerned about placing static strings in code. Even if codeclimate doesn't flag the static strings problem, it still makes the code a bit inflexible.
Another reason for proposing this change is the react application. There are two ways we can construct URLs in react apps:
1) Hardcode them
2) Create a config-like service that would provide url patterns into envUtils.ts or urls.ts
3) Place the url patterns into env.js or hardcode in urls.ts, envUtils.ts

option-1 is not preferable, option-2 doesn't exist yet and option-3 seems a better choice at the moment.

I suggested option-3, but am open to making a different choice.

[the remark this answers is not in the document]
Comment (prasadtalasila) [reply client/src/route/workbench/WorkbenchButtonData.tsx:7]: Makes sense. Please keep `envUtils.tsx` of 005fd94 commit.

[the remark this answers is not in the document]
Comment (prasadtalasila) [reply client/src/route/workbench/WorkbenchButtonData.tsx:7]: The links in config/prod.js should be:
```
Terminal --> /terminals/main
Desktop --> /tools/vnc/?password=vncpassword
VSCode --> /tools/vscode/
Jupyter Lab --> /lab
Jupyter Notebook --> '' (empty string)
```

**Comment by prasadtalasila** [comment]
@Omarley7 , may be I'm doing something very wrong.
When this last commit gets pulled to my computer, I can see the changed files, but no changes on the react site itself. The redux state persistence doesn't show up on the menu. Also the links on the workbench page don't get updated.
The commands I followed were:
```
yarn install
yarn build
yarn configapp dev
yarn start
```

**Comment by Omarley7** [comment]
@prasadtalasila That is very strange. Can you verify the following behavior?
- Redux currently only keeps the menu bar open, when navigating between routes. 
- While the app is running, change ` REACT_APP_URL_WORKBENCH: 'https://example.com/',`. Run `yarn configapp dev
`. Then refresh the page, enter email and navigate to Workbench. The tooltip should show a new base url.

**Comment by prasadtalasila** [comment]
> @prasadtalasila That is very strange.

> - While the app is running, change ` REACT_APP_URL_WORKBENCH: 'https://example.com/',`. Run `yarn configapp dev
> `. Then refresh the page, enter email and navigate to Workbench. The tooltip should show a new base url. 

This it does. But the tooltip needs to show baseurl+specific suffix for each icon. Only then can we go to pages of the respective tools.

I see that the username capture functionality seems to be in place. Do you use this username to compose relative urls?

**Comment by Omarley7** [comment]
The tooltip is showing `{workbenchurl}/{username}/{btnEndpoint}` correctly [on my end](https://ibb.co/xfBLrDC) :thinking: 
You're not even getting the username from the sign-in form?

**Comment by prasadtalasila** [comment]
> The tooltip is showing `{workbenchurl}/{username}/{btnEndpoint}` correctly [on my end](https://ibb.co/xfBLrDC) 🤔 You're not even getting the username from the sign-in form?

My mistake. The tooltip correctly shows the `https://ece.au.dk/user1/terminals/main` URL. Now I am scratching head about the discrepancy. I must have been looking at the previous commit. Sorry about the confusion.

Is this PR complete with tests? or you are going to commit more before it can be potentially merged?

**Comment by Omarley7** [comment]
I'm working on the lasts tests now. About commit unittest for the component. I'll move tests of utilities and Redux implementations to next PR

**Comment by Omarley7** [comment]
Hmm, E2E tests are still failing, I'm taking a look right now.
1 codeclimate duplication issue has been introduced, but I fail to see how to avoid it.

**Comment by Omarley7** [comment]
@prasadtalasila Should be ready to merge now.
This PR will resolve issue #42 since the test coverage is now above the threshold.
It is also now required to run `yarn configapp` before using the application, since there's no longer any default `env.js` left in `public`.

**Comment by prasadtalasila** [comment]
The code looks good. A few suggestions:
1) remove recharts package
2) rename "email address" field to "username". There is a variable called email in signin.tsx. This and it's state handler need to be renamed as well.
3) correct string in env.js: `REACT_APP_WORKBENCHLINK_VNCDESKTOP: '/tools/vnc/?password=vncpassword',` in `config/` files.
4) The `yarn configapp #prod | dev` has been made mandatory. A comment next to this command in README.md will let the developer know about the mandatory nature of this command.
5) All the tsx files have `import * as React from 'react';` as the first line. Is it possible to import only specific items and get the work done? If yes, this can be a good task for next PR.
6) The comment to `LinkButtons` function object is useful. Is the following comment written or generated? Also a correction on one of the comment text:

    ```txt
    /**
     *
     * The link is constructed by appending the user name to the end of the URL for the react site, and then appending the value of the environment variable to the end of that.
     * For example, if the URL for the react site is https://foo.com, the user name is "user1", and the value of the environment variable is "/my-workbench", then the link will be https://foo.com/user1/my-workbench.
    */
    ```

7) The code structure among envUtils, LinkIconsLib and LinkButtons is very good.
8) There seem to be no tests for src/store code. All the test coverage seems to be incidental from integration tests. If that's the case, I suggest the redux code to next PR. But, if you would like to add tests in this PR, that's also fine.

**Comment by Omarley7** [comment]
@prasadtalasila 
> 5. All the tsx files have import * as React from 'react'; as the first line. Is it possible to import only specific items and get the work done? If yes, this can be a good task for next PR.

This is a bit tricky. Normally we would use `import React from 'react'` but that gives errors with the unit tests. They will start giving errors:
```TypeError: Cannot read properties of undefined (reading 'createElement')```
If it doesn't give big performance issues, I'd like to keep it as is.

> 6. The comment to `LinkButtons` function object is useful. Is the following comment written or generated? Also a correction on one of the comment text:

This was completely autogenerated. I've cleaned it up a bit and taken your correction into account.

I've added tests for the redux store. We have another branch, where envUtil is tested, but it generates problems for the Auth tests. I'll bring that in the next PR.

**Comment by prasadtalasila** [comment]
> Code Climate has analyzed commit [f4be833](https://github.com/INTO-CPS-Association/DTaaS/commit/f4be83370adb2c4598877582ba3ccba3089b55c4) and detected **2 issues** on this pull request.
> 
> Here's the issue category breakdown:
> 
> Category	Count
> Duplication	2
> View more on [Code Climate](https://codeclimate.com/github/INTO-CPS-Association/DTaaS/pull/46).

The new .codeclimate.yml is supposed to have removed this kind of warnings. Are any modifications required there?

**Comment by Omarley7** [comment]
> The new .codeclimate.yml is supposed to have removed this kind of warnings. Are any modifications required there?

The new configuration raised [the mass](https://docs.codeclimate.com/docs/duplication#understand-the-engine) from unknown to the correct default value of `45` for typescript, which was sufficient for all previous cases.

I've just raised it to `55` for typescript.

**Review by prasadtalasila — APPROVED** [review]

**Pull request merged**